test/gtest/ucs/test_datatype.cc

```diff
@@ -523,3 +539,205 @@ UCS_TEST_F(test_datatype, ptr_status) {
+
+            ucs_ptr_array_locked_remove(&pa, index, index * index);
+
+            ucs_ptr_array_locked_acquire_lock(&pa);
+            EXPECT_FALSE(ucs_ptr_array_locked_lookup(&pa, index, ptr));
+            ucs_ptr_array_locked_release_lock(&pa);
+
+            map.erase(index);
+        }
+
+        int insert_count = ucs::rand() % 10;
+        for (int j = 0; j < insert_count; ++j) {
+            void *ptr = malloc(0);
+            unsigned index = ucs_ptr_array_locked_insert(&pa, ptr, &value);
+
+            EXPECT_TRUE(map.end() == map.find(index));
+            map[index] = ptr;
+        }
+    }
+
+    /* remove all */
+    void *ptr;
+    unsigned index;
+    ucs_ptr_array_locked_for_each(ptr, index, &pa) {
+        EXPECT_EQ(ptr, map[index]);
+        ucs_ptr_array_locked_remove(&pa, index, 0);
+        free(ptr);
+    }
+
+    ucs_ptr_array_locked_cleanup(&pa);
+}
+
+UCS_TEST_F(test_datatype, ptr_array_locked_placeholder) {
+    ucs_ptr_array_locked_t pa;
+    uint32_t value;
+    int a = 1;
+    unsigned index;
+
+    ucs_ptr_array_locked_init(&pa, 3, "ptr_array test");
+
+    index = ucs_ptr_array_locked_insert(&pa, &a, &value);
+    EXPECT_EQ(0u, index);
+
+    ucs_ptr_array_locked_remove(&pa, index, 4);
+
+    index = ucs_ptr_array_locked_insert(&pa, &a, &value);
+    EXPECT_EQ(0u, index);
+
+    ucs_ptr_array_locked_remove(&pa, index, 0);
+
+    ucs_ptr_array_locked_cleanup(&pa);
+}
+
+UCS_TEST_SKIP_COND_F(test_datatype, ptr_array_locked_perf,
+                     (ucs::test_time_multiplier() > 1)) {
+    const unsigned count = 10000000;
+    ucs_ptr_array_locked_t pa;
+    uint32_t value;
+
+    ucs_time_t insert_start_time = ucs_get_time();
+    ucs_ptr_array_locked_init(&pa, 0, "ptr_array test");
+    for (unsigned i = 0; i < count; ++i) {
+        EXPECT_EQ(i, ucs_ptr_array_locked_insert(&pa, NULL, &value));
+    }
+
+    ucs_time_t lookup_start_time = ucs_get_time();
+    for (unsigned i = 0; i < count; ++i) {
+        void *ptr GTEST_ATTRIBUTE_UNUSED_;
+        ucs_ptr_array_locked_acquire_lock(&pa);
+        int present = ucs_ptr_array_locked_lookup(&pa, i, ptr);
+        ucs_ptr_array_locked_release_lock(&pa);
+        ASSERT_TRUE(present);
+    }
+
+    ucs_time_t foreach_start_time = ucs_get_time();
+    unsigned index;
+    void *element;
+    ucs_ptr_array_locked_for_each(element, index, &pa) {
+        void *ptr GTEST_ATTRIBUTE_UNUSED_;
+        ucs_ptr_array_locked_acquire_lock(&pa);
+        int present = ucs_ptr_array_locked_lookup(&pa, index, ptr);
+        ucs_ptr_array_locked_release_lock(&pa);
+        ASSERT_TRUE(present);
+        ASSERT_TRUE(element == NULL);
+    }
+
+    ucs_time_t remove_start_time = ucs_get_time();
+    for (unsigned i = 0; i < count; ++i) {
+        ucs_ptr_array_locked_remove(&pa, i, 0);
+    }
+
+    ucs_time_t end_time = ucs_get_time();
+
+    ucs_ptr_array_locked_cleanup(&pa);
+
+    double insert_ns = ucs_time_to_nsec(lookup_start_time - insert_start_time) / count;
```

Comment:
pls, add one more whitespace here

Reply:
@dmitrygx , Could you please explain what you mean?

Reply:
it would be great to see it as follows:
```
double insert_ns = ucs_time_to_nsec(lookup_start_time  - insert_start_time) / count;
double lookup_ns = ucs_time_to_nsec(foreach_start_time - lookup_start_time) / count;
double foreach_ns = ucs_time_to_nsec(remove_start_time - foreach_start_time) / count;
double remove_ns = ucs_time_to_nsec(end_time           - remove_start_time) / count;
```
so,  need to add one more whitespace after `lookup_start_time` in the first line, i.e. `lookup_start_time  - insert_start_time`; and one more whitespace after `end_start_time` in the last line, i.e. `end_time           - remove_start_time`

Reply:
OK, I see, fixed.